R/consolidate_metadata.R

```diff
@@ -0,0 +1,125 @@
+#'   if the variable names differ across the input datasets (`datasets`).
+#'
+#'   *Permitted Values*: `"none"`, `"message"`, warning"`, `"error"`
+#'
+#' @param check_keys Check keys?
+#'
+#'   If `"warning"` or `"error"` is specified, a message is issued if the key
+#'   variables (`key_vars`) are not a unique key in all of the input datasets
+#'   (`datasets`).
+#'
+#'   *Permitted Values*: `"none"`, `"warning"`, `"error"`
+#'
+#' @author Stefan Bundfuss
+#'
+#' @details All observations of the input datasets are put together into a
+#'   single dataset. If a by group (defined by `key_vars`) exists in more than
+#'   one of the input datasets, the observation from the last dataset is
+#'   selected.
+#'
+#' @return A dataset which contains one row for each by group occuring in any of
+#'   the input datasets.
+#'
+#' @keywords create_aux
+#' @family create_aux
+#'
+#' @export
+#'
+#' @examples
+#' library(tibble)
+#' glob_ranges <- tribble(
+#'   ~PARAMCD, ~ANRLO, ~ANRHI,
+#'   "PULSE",      60,    100,
+#'   "SYSBP",      90,    130,
+#'   "DIABP",      60,     80
+#' )
+#' proj_ranges <- tribble(
+#'   ~PARAMCD, ~ANRLO, ~ANRHI,
+#'   "SYSBP",     100,    140,
+#'   "DIABP",      70,     90
+#' )
+#' stud_ranges <- tribble(
+#'   ~PARAMCD, ~ANRLO, ~ANRHI,
+#'   "BMI",        18,     25
+#' )
+#'
+#' consolidate_metadata(
+#'   datasets = list(
+#'     global = glob_ranges,
+#'     project = proj_ranges,
+#'     study = stud_ranges
+#'   ),
+#'   key_vars = vars(PARAMCD)
+#' )
+consolidate_metadata <- function(datasets,
+                                 key_vars,
+                                 source_var = SOURCE,
+                                 check_vars = "warning",
+                                 check_keys = "error") {
+  assert_list_of(datasets, class = "data.frame", named = TRUE)
+  assert_vars(key_vars)
+  source_var <- assert_symbol(enquo(source_var))
+  check_vars <-
+    assert_character_scalar(
+      check_vars,
+      values = c("none", "message", "warning", "error"),
+      case_sensitive = FALSE
+    )
+  check_keys <-
+    assert_character_scalar(
+      check_keys,
+      values = c("none", "warning", "error"),
+      case_sensitive = FALSE
+    )
+
+  if (check_vars != "none") {
+    if (length(unique(map(datasets, function(x) sort(names(x))))) > 1) {
+      msg_funs <- list(message = inform, warning = warn, error = abort)
+      msg_funs[[check_vars]](paste(
+        "The variable names differ across the input datasets.",
+        "This message can be suppressed by setting `check_vars = \"none\"`.",
+        sep = "\n"
+      ))
+    }
+  }
+
+  data_order <- seq_len(length(datasets))
+  names(data_order) <- names(datasets)
+  all_data <- bind_rows(datasets, .id = as_label(source_var))
+  tmp_source_ord <- get_new_tmp_var(all_data, prefix = "tmp_source_ord")
+  all_data %>%
+    mutate(!!tmp_source_ord := data_order[!!source_var]) %>%
+    filter_extreme(
```

Comment:
Is there ever an instance where you might want to keep a non-last observation? For example in your code if HEIGHT was in both project and study but you wanted to keep project HEIGHT over the study HEIGHT? Having the ability to do the filter step after calling `consolidate_metadata()` seems more flexible but that might be me not entirely understanding the full picture of the use case

Reply:
I am not aware of such a scenario. If you prefer project HEIGHT over study HEIGHT, why would you add HEIGHT to the study metadata at all?